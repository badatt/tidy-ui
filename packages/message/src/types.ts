import { HTMLAttributes } from 'react';
import { IBaseProps, TTone } from '@tidy-ui/types';

/**
 * Message props
 */
export interface IMessageProps extends IBaseProps, HTMLAttributes<HTMLDivElement> {
  /**
   * It `true`, creates a closable message widget with a close button at top right corner
   *
   * @type {?boolean}
   * @default false
   */
  closable?: boolean;

  /**
   * Delay timer to remove the message widget.
   * When set to 0, the message is not removed automatically
   * (in milliseconds)
   *
   * @type {?number}
   * @default 0
   */
  duration?: number;

  /**
   * Title of the message widget
   *
   * @type {?(JSX.Element | string)}
   */
  header?: JSX.Element | string;

  /**
   * Callback (if any) after the message widget is closed
   *
   * @type {?((e?: any | undefined) => void)}
   */
  // eslint-disable-next-line  @typescript-eslint/no-explicit-any
  onClose?: (e?: any | undefined) => void;

  /**
   * If `true`, creates outlined message widget, won't fill
   *
   * @type {?boolean}
   * @default false
   */
  outlined?: boolean;

  /**
   * If `true`, creates sharp corners for the message widget
   *
   * @type {?boolean}
   * @default false
   */
  sharp?: boolean;

  /**
   * If `true`, stretches the message widget to full width of the parent container
   *
   * @type {?boolean}
   * @default false
   */
  stretched?: boolean;

  /**
   * Color variant of the message, typically used to classify the
   * information based on its level
   *
   * @type {?TTone}
   */
  tone?: TTone;

  /**
   * If `true`, label wont be added to the message widget
   *
   * @type {?boolean}
   * @default false
   */
  withoutLabel?: boolean;
}
