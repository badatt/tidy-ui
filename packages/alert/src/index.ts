export * from './Alert';
export * from './AlertFooter';
export * from './AlertTitle';
