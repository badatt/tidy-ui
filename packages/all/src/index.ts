export * from '@tidy-ui/commons';
export * from '@tidy-ui/feedback';
export * from '@tidy-ui/form';
export * from '@tidy-ui/layout';
export * from '@tidy-ui/navigation';
export * from '@tidy-ui/presentation';
