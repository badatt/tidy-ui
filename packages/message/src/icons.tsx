const emojis = {
  info: '📌',
  success: '✅',
  warning: '⚠️',
  danger: '❌',
};

export { emojis };
