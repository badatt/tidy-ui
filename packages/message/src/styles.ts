import { css, hsla } from '@tidy-ui/commons';
import { IMessageProps } from './types';

/**
 * Filled label styles
 *
 * @internal
 */
const filledLabel = css<IMessageProps>`
  ${({ theme: { palette }, tone }) => css`
    background-color: ${palette[tone!][800]};
    color: ${palette[tone!][50]};
  `}
`;

/**
 * Outlined label styles
 *
 * @internal
 */
const outlinedLabel = css<IMessageProps>`
  ${({ theme: { palette, isDark }, tone }) => css`
    color: ${isDark ? palette[tone!][300] : palette[tone!][700]};
    background-color: ${isDark ? hsla(palette[tone!].shades[900], 0.4) : hsla(palette[tone!].shades[100], 0.4)};
    border: 1px solid ${isDark ? palette[tone!][500] : palette[tone!][400]};
    border-bottom: 0;
  `}
`;

/**
 * Filled content styles
 *
 * @internal
 */
const filledContent = css<IMessageProps>`
  ${({ theme: { palette }, tone }) => css`
    color: ${palette[tone!][50]};
    background-color: ${palette[tone!][700]};
  `}
`;

/**
 * Outlined content styles
 *
 * @internal
 */
const outlinedContent = css<IMessageProps>`
  ${({ theme: { palette, isDark }, tone }) => css`
    color: ${isDark ? palette[tone!][300] : palette[tone!][700]};
    border: 1px solid ${isDark ? palette[tone!][500] : palette[tone!][400]};
  `}
`;

export { filledContent, filledLabel, outlinedContent, outlinedLabel };
