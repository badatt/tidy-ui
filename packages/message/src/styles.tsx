import { css, hsla } from '@tidy-ui/theme';
import { IMessageProps } from './types';

/**
 * Filled label styles
 *
 * @internal
 */
const filledLabel = css<IMessageProps>`
  ${({ theme: { palette }, withColor }) => css`
    background-color: ${palette[withColor!][600]};
    color: ${palette[withColor!][50]};
  `}
`;

/**
 * Outlined label styles
 *
 * @internal
 */
const outlinedLabel = css<IMessageProps>`
  ${({ theme: { palette, isDark }, withColor }) => css`
    color: ${isDark ? palette[withColor!][400] : palette[withColor!][600]};
    background-color: ${isDark ? hsla(palette[withColor!].shades[800], 0.8) : palette[withColor!][100]};
    border: 1px solid ${isDark ? palette[withColor!][600] : palette[withColor!][300]};
    border-bottom: 0;
  `}
`;

/**
 * Filled content styles
 *
 * @internal
 */
const filledContent = css<IMessageProps>`
  ${({ theme: { palette }, withColor }) => css`
    background-color: ${palette[withColor!][500]};
    color: ${palette[withColor!][50]};
  `}
`;

/**
 * Outlined content styles
 *
 * @internal
 */
const outlinedContent = css<IMessageProps>`
  ${({ theme: { palette, isDark }, withColor }) => css`
    color: ${isDark ? palette[withColor!][400] : palette[withColor!][600]};
    background-color: ${isDark ? hsla(palette[withColor!].shades[900], 0.3) : palette[withColor!][50]};
    border: 1px solid ${isDark ? palette[withColor!][600] : palette[withColor!][300]};
  `}
`;

export { filledContent, filledLabel, outlinedContent, outlinedLabel };
