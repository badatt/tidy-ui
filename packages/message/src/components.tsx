import { css, styled } from '@tidy-ui/commons';
import { IPalette } from '@tidy-ui/types';
import { filledContent, filledLabel, outlinedContent, outlinedLabel } from './styles';
import { IMessageProps } from './types';

/**
 * Internal Message root component
 *
 * @internal
 */
const MessageRoot = styled.div<IMessageProps>`
  ${({ stretched, margin }) => css`
    margin: ${margin};
    width: ${stretched ? '100%' : 'fit-content'};
    position: relative;
  `}
`;

/**
 * Close button color
 *
 * @param {boolean} isDark isDark check
 * @param {IPalette} palette theme palette
 * @param {string} color color string
 * @returns {string} color in hsla
 */
const closeButtonColor = (isDark: boolean, palette: IPalette, color: string): string => {
  return isDark ? palette[color][400] : palette[color][600];
};

/**
 * CloseButton component
 *
 * @internal
 */
const CloseButton = styled.i<IMessageProps>`
  ${({ theme: { palette, isDark }, tone, outlined }) => css`
    height: 1rem;
    width: 1rem;
    position: absolute;
    color: ${palette[tone!][50]};
    cursor: pointer;
    ${outlined
      ? css`
          top: 26px;
          right: 2px;
          color: ${closeButtonColor(isDark, palette, tone!)};
        `
      : css`
          top: 24px;
          right: 2px;
          color: ${palette[tone!][50]};
        `}
  `}
`;

/**
 * Message widget label component
 *
 * @internal
 */
const MessageLabel = styled.div<IMessageProps>`
  ${({ theme: { layout, typography }, sharp, outlined }) => css`
    display: flex;
    justify-content: center;
    align-items: center;
    gap: 2px;
    border-top-left-radius: ${!sharp && layout.radius};
    border-top-right-radius: ${!sharp && layout.radius};
    width: fit-content;
    padding: 0.25rem 1rem;
    font-size: 0.625rem;
    text-transform: uppercase;
    font-weight: ${typography.fontWeightBold};
    ${outlined ? outlinedLabel : filledLabel}
    * {
      height: 0.75rem;
      width: 0.75rem;
    }
  `}
`;

/**
 * MessageContent component
 *
 * @internal
 */
const MessageContent = styled.div<IMessageProps>`
  ${({ theme: { layout }, sharp, stretched, withoutLabel, outlined }) => css`
    padding: 0.75rem 1rem;
    border-radius: ${!sharp && layout.radius};
    ${!withoutLabel &&
    css`
      border-top-left-radius: 0;
    `}
    width: ${stretched ? '100%' : 'fit-content'};
    ${outlined ? outlinedContent : filledContent}
  `}
`;

export { CloseButton, MessageContent, MessageLabel, MessageRoot };
