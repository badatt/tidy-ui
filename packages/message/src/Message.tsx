import React from 'react';
import { Icon, TonedIcon, useIsMounted } from '@tidy-ui/commons';
import { CloseButton, MessageContent, MessageLabel, MessageRoot } from './components';
import { IMessageProps } from './types';

/**
 * Message component can be used to display important information on a page based on the
 * information importance level, powered by various styles (with theming) and settings
 *
 */
const Message = React.forwardRef<HTMLDivElement, IMessageProps>((props, ref) => {
  const { children, className, closable, ele, onClose, ...rest } = props;

  const isMounted = useIsMounted();
  const [isHidden, setHidden] = React.useState<boolean>(false);

  const handleClose = React.useCallback(
    (e: React.MouseEvent<HTMLElement>) => {
      onClose?.(e);
      if (isMounted()) {
        setHidden(true);
      }
    },
    [isMounted, onClose],
  );

  if (isHidden) {
    return null;
  }

  return (
    <MessageRoot className={className} ref={ref} role="note" {...rest}>
      {!props.withoutLabel && (
        <MessageLabel {...rest}>
          <TonedIcon status={rest.tone} />
          {rest.tone}
        </MessageLabel>
      )}
      <MessageContent {...rest}>
        {ele ? React.cloneElement(ele, {}, children) : children}
        {closable && (
          <CloseButton
            onClick={handleClose}
            tone={rest.tone}
            outlined={rest.outlined}
            data-test-id="close-button"
            role="button"
          >
            <Icon.Close />
          </CloseButton>
        )}
      </MessageContent>
    </MessageRoot>
  );
});

Message.defaultProps = {
  closable: false,
  outlined: false,
  sharp: false,
  stretched: false,
  tone: 'info',
  withoutLabel: false,
};

Message.displayName = 'Message';

export { Message };
