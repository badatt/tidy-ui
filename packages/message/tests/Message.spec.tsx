/**
 * @jest-environment jsdom
 */
import React from 'react';
import { act } from 'react-dom/test-utils';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import { orchidDark, orchidLight, TidyUiProvider } from '../../commons/src';
import { Tone } from '../../commons/src';
import { Message } from '../src';

const text = `Lorem ipsum dolor sit, amet consectetur adipisicing elit. Mollitia nostrum voluptatibus distinctio at sapiente
      ratione blanditiis cum quam fugit commodi odit, hic eius itaque exercitationem vitae voluptates modi accusamus
      eaque!`;

describe('Message', () => {
  it('Basic render', () => {
    const tree = render(
      <TidyUiProvider theme={orchidLight}>
        <Message>{text}</Message>
      </TidyUiProvider>,
    );
    expect(tree).toMatchSnapshot();
  });
  it('Basic render in dark mode', () => {
    const tree = render(
      <TidyUiProvider theme={orchidDark}>
        <Message>{text}</Message>
      </TidyUiProvider>,
    );
    expect(tree).toMatchSnapshot();
  });
  it('Message with all color variants', () => {
    const tree = render(
      <TidyUiProvider theme={orchidLight}>
        {Object.values(Tone)
          .filter((i) => !isNaN(Number(i)))
          .map((c) => (
            <Message tone={Tone[c]} key={c}>
              {text}
            </Message>
          ))}
      </TidyUiProvider>,
    );
    expect(tree).toMatchSnapshot();
  });
  it('Message in dark mode with all color variants', () => {
    const tree = render(
      <TidyUiProvider theme={orchidDark}>
        {Object.values(Tone)
          .filter((i) => !isNaN(Number(i)))
          .map((c, i) => (
            <Message tone={Tone[c]} key={c}>
              {text}
            </Message>
          ))}
      </TidyUiProvider>,
    );
    expect(tree).toMatchSnapshot();
  });
  it('Message with all color filled variants', () => {
    const tree = render(
      <TidyUiProvider theme={orchidLight}>
        {Object.values(Tone)
          .filter((i) => !isNaN(Number(i)))
          .map((c) => (
            <Message tone={Tone[c]} key={c} isFilled>
              {text}
            </Message>
          ))}
      </TidyUiProvider>,
    );
    expect(tree).toMatchSnapshot();
  });
  it('Message with all color filled variants dark mode', () => {
    const tree = render(
      <TidyUiProvider theme={orchidDark}>
        {Object.values(Tone)
          .filter((i) => !isNaN(Number(i)))
          .map((c) => (
            <Message tone={Tone[c]} key={c} isFilled>
              {text}
            </Message>
          ))}
      </TidyUiProvider>,
    );
    expect(tree).toMatchSnapshot();
  });
  it('Message fit to content only', () => {
    const tree = render(
      <TidyUiProvider theme={orchidLight}>
        <Message isStretched={false}>{text}</Message>
      </TidyUiProvider>,
    );
    expect(tree).toMatchSnapshot();
  });
  it('Message with close button', () => {
    const tree = render(
      <TidyUiProvider theme={orchidLight}>
        <Message closable>{text}</Message>
      </TidyUiProvider>,
    );
    expect(tree).toMatchSnapshot();
  });
  it('Message with custom element', () => {
    const tree = render(
      <TidyUiProvider theme={orchidLight}>
        <Message ele={<div style={{ display: 'flex' }} />} label="Label">
          {text}
        </Message>
      </TidyUiProvider>,
    );
    expect(tree).toMatchSnapshot();
  });
  it('Message with sharp label', () => {
    const tree = render(
      <TidyUiProvider theme={orchidLight}>
        <Message isSharp>{text}</Message>
      </TidyUiProvider>,
    );
    expect(tree).toMatchSnapshot();
  });
  it('Message with no label', () => {
    const tree = render(
      <TidyUiProvider theme={orchidLight}>
        <Message hasLabel={false}>{text}</Message>
      </TidyUiProvider>,
    );
    expect(tree).toMatchSnapshot();
  });

  it('Closable message without any callback', () => {
    const tree = render(
      <TidyUiProvider theme={orchidDark}>
        <Message closable>{text}</Message>
      </TidyUiProvider>,
    );
    expect(tree).toMatchSnapshot();
    act(() => {
      const closeButton = screen.getByRole('button', { hidden: true });
      fireEvent.click(closeButton);
    });
  });

  it('Closable message without any callback filled', () => {
    const tree = render(
      <TidyUiProvider theme={orchidDark}>
        <Message closable isFilled>
          {text}
        </Message>
      </TidyUiProvider>,
    );
    expect(tree).toMatchSnapshot();
    act(() => {
      const closeButton = screen.getByRole('button', { hidden: true });
      fireEvent.click(closeButton);
    });
  });

  it('Closable message with an onClose callback', () => {
    const mockCallback = jest.fn();
    const tree = render(
      <TidyUiProvider theme={orchidLight}>
        <Message closable onClose={mockCallback}>
          {text}
        </Message>
      </TidyUiProvider>,
    );
    expect(tree).toMatchSnapshot();
    act(() => {
      const closeButton = screen.getByRole('button', { hidden: true });
      fireEvent.click(closeButton);
    });
    expect(mockCallback).toBeCalled();
  });
});
