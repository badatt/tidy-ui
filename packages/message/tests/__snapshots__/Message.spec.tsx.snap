// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message Default render 1`] = `
<div
  className=""
  color="info"
  role="message"
>
  <div
    className=""
    color="info"
  >
    <svg
      aria-hidden="true"
      fill="currentColor"
      focusable={false}
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
      />
    </svg>
    info
  </div>
  <div
    className=""
    color="info"
  />
</div>
`;
